crates/wac-graph/src/graph.rs

```diff
@@ -0,0 +1,1884 @@
+    ///
+    /// Returns `None` if the given id is invalid or if the node is not an alias.
+    pub fn get_alias_source(&self, node: NodeId) -> Option<(NodeId, &str)> {
+        for (s, t, edge) in self.graph.edges_directed(node, Direction::Incoming) {
+            assert_eq!(t, node);
+
+            if let Edge::Alias(index) = edge {
+                let data = self.node_data(s);
+                match data.item_kind {
+                    ItemKind::Instance(id) => {
+                        let types = self.node_types(data);
+                        let export = types[id].exports.get_index(*index).unwrap().0;
+                        return Some((s, export));
+                    }
+                    _ => unreachable!("alias source should be an instance"),
+                }
+            }
+        }
+
+        None
+    }
+
+    /// Gets the satisfied arguments of an instantiation node.
+    ///
+    /// Returns an iterator over the argument names and the node used to satisfy the argument.
+    ///
+    /// If the node identifier is invalid or if the node is not an instantiation node, an
+    /// empty iterator is returned.
+    pub fn get_arguments(&self, node: NodeId) -> impl Iterator<Item = (&str, NodeId)> {
+        self.graph
+            .edges_directed(node, Direction::Incoming)
+            .filter_map(|(s, t, e)| {
+                let target = self.node_data(t);
+                let imports = match target.kind {
+                    NodeKind::Instantiation(_) => {
+                        let package = &self.packages[target.package?.index].package.as_ref()?;
+                        &package.types()[package.ty()].imports
+                    }
+                    _ => return None,
+                };
+
+                match e {
+                    Edge::Alias(_) => None,
+                    Edge::Argument(indexmap) => Some(
+                        indexmap
+                            .iter()
+                            .map(move |i| (imports.get_index(*i).unwrap().0.as_ref(), s)),
+                    ),
+                }
+            })
+            .flatten()
+    }
+
+    /// Gets information about a node in the graph.
+    ///
+    /// Returns `None` if the specified node identifier is invalid.
+    pub fn get_node_info(&self, id: NodeId) -> Option<NodeInfo<C>> {
+        self.get_node(id)?;
+        let data = self.node_data(id);
+
+        Some(NodeInfo {
+            types: self.node_types(data),
+            kind: data.item_kind,
+            context: &data.context,
+            name: data.name.as_deref(),
+            aliases: &data.aliases,
+            export: data.export.as_deref(),
+        })
+    }
+
+    /// Sets the name of a node in the graph.
+    ///
+    /// Node names are recorded in a `names` custom section when the graph is encoded.
+    pub fn set_node_name(&mut self, id: NodeId, name: impl Into<String>) -> GraphResult<()> {
+        self.get_node(id).ok_or(Error::InvalidPackageId)?;
```

Comment:
The error here and in `export_node` is incorrect (these aren't package ids).

Reply:
Whoops, good catch!